Major Logic Update

Eternity-Lotus opened the pull request:
Added HARD and MINOR GLITCHES logic to certain checks, fixed a couple bugs associated with a couple of existing checks.

Eternity-Lotus committed:
Update Main.java

Added HARD and MINOR GLITCHES logic to certain checks, fixed a couple bugs associated with a couple existing checks.

Eternity-Lotus committed:
Update Main.java

Fixed a bug in W3S that was causing it to return unexpectedly. Added HARD and MINOR GLITCHES logic, implemented MERCILESS difficulty.

Eternity-Lotus committed:
Update Difficulty.java

MERCILESS Difficulty added

Eternity-Lotus committed:
Added MERCILESS Difficulty as an option to the GUI

AaronDobbe commented on src/GUI.java:
```diff
@@ -67,13 +67,14 @@ public GUI() {
                 "<b>Hard</b>: A hardcore difficulty that requires speedrun-level tricks and very difficult/odd maneuvers, but no glitches.<br>" +
-                "<b>Hard + Minor Glitches</b>: A super-hardcore difficulty that also requires wallclips, but no out-of-bounds or screenwrapping.</html>");
+                "<b>Hard + Minor Glitches</b>: A super-hardcore difficulty that also requires wallclips, but no out-of-bounds or screenwrapping.<br>" +
+				"<b>Merciless: An extremely hardcore difficulty that also requires screen wraps, wrong warps, and double boosts.</html>");
```
This needs a closing tag `</b>` after the difficulty name

AaronDobbe commented on src/GUI.java:
You'll also need to add `"merciless"` (matching the case you added to Main.java) to the other `logicOptions` array, which is defined inside `actionPerformed()` (line 293, I believe). That one defines the strings that actually get passed to the seed generation process

Eternity-Lotus replied:
Added!

AaronDobbe commented on src/Main.java:
```diff
@@ -1511,10 +1530,14 @@ else if (location.equals("N5B")) {
         else if (location.equals("N6S")) {
+			/*
+			* New logic addition for this chest for MINOR GLITCHES. 
+			* You can skip the need for Boots using a Walljump when compared to HARD.
```
A question - is this an off-screen walljump? I've been reluctant to include those in MG difficulty, for the sake of people who use visual cues to hit the timing

Eternity-Lotus replied:
This is not an off-screen walljump. The only difference in regards to this walljump in comparison is that you're jumping into solid blocks, thus you're needing Garlic and Helmet to execute it. Other than that, it's no different from other such walljumps.

AaronDobbe commented on src/Main.java:
```diff
@@ -1585,22 +1620,45 @@ else if (location.equals("W3G")) {
         }
         else if (location.equals("W3B")) {
+			/*
+			* Added MERCILESS logic to this chest.
```
This is sick, great suggestion for Merciless

Does it really not require flippers? (I believe you if that's correct, I've just never seen this before)

Eternity-Lotus replied:
This is correct. Once you walk off the top pipe, if you've done everything correct up to that point, you will warp to the bottom of the pool with the block lifted up as if you had the air pump. However, it's treated as if you're still on the surface, thus you get kind of mixed properties on it. Since you're right on the door tiles after the warp, you can press Up again and you will enter the Blue Chest room.

AaronDobbe commented on src/Main.java:
```diff
@@ -1585,22 +1620,45 @@ else if (location.equals("W3G")) {
-                    && inventory.contains(Items.PUMP)
-                    && canSwim(inventory);
+                    && (inventory.contains(Items.PUMP) && canSwim(inventory))
+					|| (difficulty >= Difficulty.MERCILESS && canGP(inventory) && canLift(inventory) && inventory.contains(item.JUMP_BOOTS));
```
`item.JUMP_BOOTS` should be `Items.JUMP_BOOTS` here

Eternity-Lotus replied:
Will get this fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -1585,22 +1620,45 @@ else if (location.equals("W3G")) {
-                    && (inventory.contains(Items.SPIKED_HELMET) || (difficulty > Difficulty.EASY && canLift(inventory) && inventory.contains(Items.JUMP_BOOTS)));
+                    && (inventory.contains(Items.SPIKED_HELMET) 
+						|| (difficulty > Difficulty.EASY && canLift(inventory) && inventory.contains(Items.JUMP_BOOTS))
+						|| (difficulty >= Difficulty.S_HARD && inventory.contains(Items.JUMP_BOOTS));
```
This needs one more `)` before the `;`

(to match the `(` at the start of line 1648)

Eternity-Lotus replied:
Fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -1771,13 +1837,22 @@ else if (location.equals("E1R")) {
-                    && inventory.contains(Items.JUMP_BOOTS)
-                    && inventory.contains(Items.STONE_FOOT);
+                    && inventory.contains(Items.STONE_FOOT)
+					&& (inventory.contains(Items.JUMP_BOOTS) || Difficulty = Difficulty.S_HARD);
```
`Difficulty = Difficulty.S_HARD` should be `difficulty >= Difficulty.S_HARD`

Eternity-Lotus replied:
Fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -1771,13 +1837,22 @@ else if (location.equals("E1R")) {
             return canAccess("E1", inventory)
-                    && inventory.contains(Items.DETONATOR);
+                && ((inventory.contains(Items.DETONATOR))
+					|| (canLift(inventory) && inventory.contains(ITEMS.JUMP_BOOTS) && inventory.contains(ITEMS.SPIKED_HELMET)));
```
This needs a difficulty check, and both instances of `ITEMS` should be `Items`

Eternity-Lotus replied:
Fixed all. Looks like I had forgotten to actually add the difficulty part itself even though I had it commented out.

AaronDobbe commented on src/Main.java:
```diff
@@ -2079,7 +2183,13 @@ else if (region == 0x1d) {
+				* This only works at night since the water level is higher at night.
+				*/
+                    return (inventory.contains(Items.GARLIC)
+						|| (difficulty >= Difficulty.MERCILESS && canSwim(inventory)));
```
If this only works at night, it should also check for `!daytime`

Eternity-Lotus replied:
`!daytime` has been added.

AaronDobbe commented on src/Main.java:
```diff
@@ -2139,11 +2249,18 @@ else if (location == 3) {
             }
             else if (region == 0x5) {
+			/*
+			* New addition for this check for MINOR GLITCHES. 
```
Same question as the N6S chest check, about whether this is a offscreen walljump

Eternity-Lotus replied:
It is not. Only difference with this walljump when compared to others is that you're needing Garlic and Helmet to break the solid blocks necessary. The screen does scroll if you're standing up there, but other than that it's not really any different than other walljumps.

AaronDobbe commented on src/Main.java:
```diff
@@ -2139,11 +2249,18 @@ else if (location == 3) {
+			*/
                 return inventory.contains(Items.GARLIC) && inventory.contains(Items.SPIKED_HELMET) && canGP(inventory)
-                        && (canSwim(inventory)
+                        && canSwim(inventory)
```
There's an extra `;` in here, but besides that this isn't quite right; I'd suggest keeping the original structure of this check, but just replacing `inventory.contains(Items.JUMP_BOOTS)` with `(inventory.contains(Items.JUMP_BOOTS) || difficulty >= Difficulty.S_HARD)`

Eternity-Lotus replied:
Removed extra semicolon and restructured to 
`return inventory.contains(Items.GARLIC) && inventory.contains(Items.SPIKED_HELMET) && canGP(inventory)
                        && canSwim(inventory)
                            || (difficulty >= Difficulty.HARD
                                && (inventory.contains(Items.JUMP_BOOTS) || difficulty >= Difficulty.S_HARD));`

AaronDobbe commented on src/Main.java:
```diff
@@ -2257,7 +2374,21 @@ else if (level.equals("W3")) {
+					*/
+                    return canGP(inventory)
+						|| (difficulty >= Difficulty.HARD
+							&& (canSwim(inventory) || (inventory.contains(Items.BEANSTALK_SEEDS) && (keyColor == 0 || keyColor == 1)))					keyColor == 1)
```
Something's up with this line, it looks like you had something pasted here or scrolled too far to the right

Eternity-Lotus replied:
I did use copy and paste yes, but I don't know what happened there. I removed the extra part. I actually also had to move one of the `)` because it was actually in the wrong spot. Hard logic needs to check for all of these, it was previously only checking the first one.

AaronDobbe commented on src/Main.java:
```diff
@@ -2281,17 +2412,47 @@ else if (region == 0x14) {
+				* Note, a Soft Reset is required. Do this soft reset after making a suspend save before the Throw + Dashjump wallclip. 
+				* This execution skips the need for the Air Pump.
+				*/
+                return (inventory.contains(Items.PUMP) && canSwim(inventory)
```
This line needs another `)` after the `canSwim()` check

Eternity-Lotus replied:
Yup, indeed that is the case. Fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -2281,17 +2412,47 @@ else if (region == 0x14) {
+				* This execution skips the need for the Air Pump.
+				*/
+                return (inventory.contains(Items.PUMP) && canSwim(inventory)
+					|| (difficulty >= Difficulty.MERCILESS && canGP(inventory) && canLift(inventory) && inventory.contains(item.JUMP_BOOTS));
```
`item.JUMP_BOOTS` should be `Items.JUMP_BOOTS`

Eternity-Lotus replied:
Also fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -2281,17 +2412,47 @@ else if (region == 0x14) {
+					* You can use Ladder Scrolling to skip the need for the Golden Glove and the Boots 
+					* when compared to MINOR GLITCHES.
+					*/
+					return (difficulty >= Difficulty.MERCILESS || canSuperLift(inventory))
```
I'd wrap the entire first half of this in a set of `()` so that, at the top level, it's `(first method) || (second method)`

Eternity-Lotus replied:
Done. Also removed the extra `)` at the end of the 2nd method here as a result.

AaronDobbe commented on src/Main.java:
```diff
@@ -2281,17 +2412,47 @@ else if (region == 0x14) {
+					*/
+					return (difficulty >= Difficulty.MERCILESS || canSuperLift(inventory))
+						&& (difficulty >= Difficulty.S_HARD || inventory.contains(Items.JUMP_BOOTS))
+						&& canSuperGP(inventory);
```
This `;` needs to be deleted

Eternity-Lotus replied:
Removed

AaronDobbe commented on src/Main.java:
```diff
@@ -2281,17 +2412,47 @@ else if (region == 0x14) {
+							|| (difficulty >= Difficulty.S_HARD 
+								&& inventory.contains(Items.JUMP_BOOTS) 
+								&& inventory.contains(Items.SPIKED_HELMET)
+								&& (difficulty >= Difficulty.MERCILESS || canSuperLift(inventory))));
```
There's one too many `)` at the end here

Eternity-Lotus replied:
(Fixed above)

AaronDobbe commented on src/Main.java:
```diff
@@ -2281,17 +2412,47 @@ else if (region == 0x14) {
-                return (difficulty >= Difficulty.S_HARD || inventory.contains(Items.JUMP_BOOTS))
-                        && canSuperGP(inventory)
-                        && canSuperLift(inventory);
+            else if (region == 0x5) 
```
This line needs a `{`

Eternity-Lotus replied:
Added!

AaronDobbe commented on src/Main.java:
```diff
@@ -2487,13 +2683,19 @@ else if (region == 0x14) {
                     return canSwim(inventory)
-                            && inventory.contains(Items.GARLIC);
+                        && (inventory.contains(Items.GARLIC)
+						|| (difficulty >= Difficulty.S_HARD && inventory.constructor(Items.JUMP_BOOTS));
```
`constructor` should be `contains`

Eternity-Lotus replied:
Looks like I must have tabbed too early on this one so it gave the wrong function. Fixed

AaronDobbe commented on src/Main.java:
```diff
@@ -2567,7 +2769,11 @@ else if (region == 0x7) {
+					* Added MERCILESS logic for this check, which is Main Area - Lower Right.
+					* Using Ladder Scrolling at the start, this check can be reached without a Glove.
+					*/
+                    return (canLift(inventory) || difficulty >= difficulty.MERCILESS);
```
`difficulty.MERCILESS` should be `Difficulty.MERCILESS`

Eternity-Lotus replied:
This will be fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -2600,7 +2808,13 @@ else if (region == 0x5) {
+					* Note, a soft reset is required for this to work, and this wrong warp only works during the day.
+					* The soft reset can occur either immediately before entering the stage or after a suspend save.
+					*/
+                    return (inventory.contains(Items.SCISSORS) || (dayOnly && difficulty >= Difficulty.MERCILESS));
```
`dayOnly` here should be `daytime`

Eternity-Lotus replied:
Will get this fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -2657,10 +2871,25 @@ else if (level.equals("E2")) {
+					* Note, make sure you're holding neutral when the crouch charge ends.
+					*/
+                    return daytime || canSuperSwim(inventory)
+						|| (difficulty >= Difficulty.MERCILESS && canSwim(inventory) && inventory.contains(ITEMS.GARLIC));
```
`ITEMS` should be `Items`

Eternity-Lotus replied:
Will get this fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -2676,7 +2905,17 @@ else if (region == 0x15) {
+					* Finally, jump up to get past the current and climb the ladder to reach this region without dayTime or Super Flippers.
+				*/
+                return daytime || canSuperSwim(inventory)
+					|| (difficulty >= Difficulty.MERCILESS && canSwim(inventory) && inventory.contains(ITEMS.GARLIC));
```
Same as above, `ITEMS` should be `Items`

Eternity-Lotus replied:
Fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -2840,7 +3092,15 @@ else if (region == 0x4) {
+				* Stun it, then walk into it towards the left and knock it down the ledges necessary.
+				* Once it's off the ledge that the door leading to the Hammerbot room is on, do the jumps across like normal to reach the check.
+				*/
+                    return (difficulty > Difficulty.EASY && inventory.contains(Items.VALVE) && canLift(inventory) && inventory.contains(Items.JUMP_BOOTS)
```
This is missing a `)` somewhere, but I'd restructure this anyway - better to just have it the way it was before, but replace `canLift(inventory)` with `(difficulty >= Difficulty.S_HARD || canLift(inventory))`

Eternity-Lotus replied:
This will be restructured to
`
return difficulty > Difficulty.EASY 
	&& inventory.contains(Items.VALVE) 
	&& inventory.contains(Items.JUMP_BOOTS)
        && (difficulty >= Difficulty.S_HARD || canLift(inventory));
`

AaronDobbe reviewed (changes requested):
This is mostly good as far as I can tell! Added some comments on stuff that needs to be changed, but it's mostly just typos and the occasional missing symbol. Once these are fixed I'll be able to build the executable with your changes and make sure nothing's immediately broken.

I appreciate this a lot, thank you for your help making this randomizer the best it can be :)

Eternity-Lotus committed:
Added Changes made from Pull Request Comments

Eternity-Lotus committed:
Updates to syntax, More Logic Updates added

Comments in the Pull Request handled, Logic updates added in S5

Eternity-Lotus committed:
Added MINOR GLITCHES logic to S4

Eternity-Lotus committed:
Added MERCILESS logic check for E2 Blue

The checks necessary were already checking for MERCILESS logic, just not the chest itself.

Eternity-Lotus committed:
Added MINOR GLITCHES Logic for W3S Chest

Eternity-Lotus committed:
Added MINOR GLITCHES logic for N4 Blue Areas

Eternity-Lotus committed:
Added Flippers/Blue Key Check to W3 Merciless Logic

The Jellyfish Room in W3 now checks if the region contains specifically the Blue Key, or if the player has Flippers. Just like with the Spark Room in N6, it's impossible to escape once the player exits the water. Therefore, any Key Color besides the Blue will logically require Flippers to obtain.

Eternity-Lotus committed:
W3 Jellyfish Room logic update

Tested this again. It is possible to escape using the same water treading trick in N6. The block will be down, thus blocking the door again, but you can get out of the block by simply moving right. As such, this logic has now been updated so that it checks for any key color besides the Green. The Green Key will still logically require Flippers, but that should be the only exception.

Eternity-Lotus committed:
Added Minor Glitches Logic to E3 Spike Maze

AaronDobbe commented on src/Main.java:
```diff
@@ -1506,8 +1506,12 @@ else if (location.equals("N4G")) {
+			*/
             return canAccess("N4", inventory)
-                    && inventory.contains(Items.PUMP)
+                    && ((difficulty >= Difficulty.S_HARD && canSwim(inventory) && inventory.contains(Items.JUMP_BOOTS)) || inventory.contains(Items.PUMP))
```
This looks good, and I imagine isn't very hard to execute - it might be nice to have pixel lineups documented for this somewhere

Eternity-Lotus replied:
Sure, I can send you a sample pixel alignment that works for the charge while attempting to make this clip. That said, it's worth noting that waterclips are actually quite a bit more forgiving than normal walljumps. There tends to be multiple pixel sets that work for them (many more than walljumps and reverse walljumps). This one is only a little more finicky because of the fact that you need to do a high jump to succeed the clip.

AaronDobbe commented on src/Main.java:
```diff
@@ -1603,13 +1607,15 @@ else if (location.equals("W3S")) {
-                        && canLift(inventory)
-                        && inventory.contains(Items.JUMP_BOOTS)));
+						&& inventory.contains(Items.JUMP_BOOTS)
+                        && (difficulty >= Difficulty.S_HARD || canLift(inventory)));
```
This needs one more `)` before the semicolon, otherwise looks good

Eternity-Lotus replied:
Yup, looks like I missed this one. Fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -1840,7 +1858,7 @@ else if (location.equals("E1G")) {
             return canAccess("E1", inventory)
                     && inventory.contains(Items.STONE_FOOT)
-					&& (inventory.contains(Items.JUMP_BOOTS) || Difficulty = Difficulty.S_HARD);
+					&& (inventory.contains(Items.JUMP_BOOTS) || difficulty = Difficulty.S_HARD);
```
The `=` here still needs to be replaced with `>=` (both because this should apply to Merciless as well, and because a single `=` is assignment rather than comparison. Thankfully, this wouldn't compile because it's part of a boolean `||`, but if it did... the randomizer would actually change the difficulty of the seed to Minor Glitches every time this logic was checked!)

Eternity-Lotus replied:
That's a good point, don't know why I overlooked this one after going through this many times. Fixed!

AaronDobbe commented on src/Main.java:
```diff
@@ -1867,8 +1885,18 @@ else if (location.equals("E2G")) {
+			*/
             return canAccess("E2", inventory)
-                    && (dayOnly || canSuperSwim(inventory));
+                    && (dayOnly || canSuperSwim(inventory))
+					|| (difficulty >= Difficulty.MERCILESS && canSwim(inventory) && inventory.contains(Items.GARLIC));;
```
The parentheses here need adjusting - I'd suggest moving the `)` at the end of line 1898 to the end of 1899, just before the semicolon. The way it is now, due to the order of operations, I believe this could be considered in logic on Merciless difficulty even if you can't access E2!

Also, there's an extra `;` at the end of 1899.

Eternity-Lotus replied:
Yup, that definitely would've been a problem. Fixed this, and removed the extra semicolon.

AaronDobbe commented on src/Main.java:
```diff
@@ -2386,9 +2423,9 @@ else if (location == 1) {
                     return canGP(inventory)
 						|| (difficulty >= Difficulty.HARD
-							&& (canSwim(inventory) || (inventory.contains(Items.BEANSTALK_SEEDS) && (keyColor == 0 || keyColor == 1)))					keyColor == 1)
+							&& (canSwim(inventory) || (inventory.contains(Items.BEANSTALK_SEEDS) && (keyColor == 0 || keyColor == 1))
```
Looking at this now that the syntax is fixed, I'm not sure I'm understanding the need to restrict the key color here; the only reason it should be necessary to restrict key colors is if your access to a key location permanently cuts you off from a chest you could otherwise reach (e.g. getting locked into the gray chest room in E6 because you don't have super overalls to get out). Is there a reason why doing this prevents you from exploring the level as you could before? Sure, you can't reach the green/blue chests without flippers anyway, but in that case the randomizer wouldn't even be trying to place the green or blue keys.

Eternity-Lotus replied:
I'll ask you about this one on disc.

Eternity-Lotus replied:
Resolved on disc, code was updated to remove the key checks here.

AaronDobbe commented on src/Main.java:
```diff
@@ -2422,24 +2459,25 @@ else if (region == 0x1a) {
-					|| (difficulty >= Difficulty.MERCILESS && canGP(inventory) && canLift(inventory) && inventory.contains(item.JUMP_BOOTS));
+                return (inventory.contains(Items.PUMP) && canSwim(inventory))
+					|| (difficulty >= Difficulty.MERCILESS && canGP(inventory) && canLift(inventory) && inventory.contains(Items.JUMP_BOOTS) 
+						&& (canSwim(inventory) || keyColor != 2));
```
This one does need a key color restriction, good catch. We don't want a situation where the player is required to, e.g., get the red key and be expected to somehow leave without flippers. That said, this is the blue area, right? The key color should be compared to `3` instead of `2` in that case (they follow the standard in-game order, `SRGB`, starting at `0`)

Eternity-Lotus replied:
I'll ask about this one on disc too due to the same reason as the added HARD logic in W3.

Eternity-Lotus replied:
Resolved on disc, code was updated to check specifically for the Blue Key with no Flippers, otherwise Flippers logically required.

AaronDobbe commented on src/Main.java:
```diff
@@ -2695,7 +2733,7 @@ else if (region == 0x19) {
                     return canSwim(inventory)
                         && (inventory.contains(Items.GARLIC)
-						|| (difficulty >= Difficulty.S_HARD && inventory.constructor(Items.JUMP_BOOTS));
+						|| (difficulty >= Difficulty.S_HARD && inventory.contains(Items.JUMP_BOOTS));
```
Must have missed this last time, but this also needs another `)` before the `;`

Eternity-Lotus replied:
Yup, that's a good catch. Will be fixed in the next commit!

AaronDobbe commented on src/Main.java:
```diff
@@ -2739,30 +2777,62 @@ else if (level.equals("S4")) {
+				else {
+					return inventory.contains(Items.RUST_SPRAY) 
+						&& inventory.contains(Items.JUMP_BOOTS)
+                        && (canLift(inventory)
```
This needs a `;` at the end, and there's an extra `(` before `canLift()`

Eternity-Lotus replied:
Whoops! Fixed.

AaronDobbe reviewed (changes requested):
Sorry for the wait! Again, this is mostly great stuff, just a few small changes requested. I really appreciate your hard work on this, and the next round of review shouldn't take me as long to get to. Thanks again!!

Eternity-Lotus committed:
Bug fixes, redid the key checks in W3 Logic Changes

Eternity-Lotus committed:
Added MINOR GLITCHES Logic for Reaching the Red Chest in S6

AaronDobbe commented on src/Main.java:
```diff
@@ -2811,7 +2811,7 @@ else if (level.equals("S5")) {
 					return inventory.contains(Items.RUST_SPRAY) 
 						&& inventory.contains(Items.JUMP_BOOTS)
-                        && (canLift(inventory)
+                        && (canLift(inventory);
```
The extra `(` at the start of this line still needs to be removed

Eternity-Lotus replied:
Done!

AaronDobbe reviewed (changes requested):
Just one last minor thing that needs to be addressed, then I think this will be good to merge in! Thank you!

Eternity-Lotus committed:
Bug fix, added logic to S6

AaronDobbe commented:
Looks great, merged! Thanks again for your contribution, and I'll work on getting the next release candidate ready to go ASAP!
